worker-thread.go

```diff
@@ -0,0 +1,255 @@
+package frankenphp
+
+// #include "frankenphp.h"
+import "C"
+import (
+	"fmt"
+	"net/http"
+	"path/filepath"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// representation of a thread assigned to a worker script
+// executes the PHP worker script in a loop
+// implements the threadHandler interface
+type workerThread struct {
+	state         *threadState
+	thread        *phpThread
+	worker        *worker
+	fakeRequest   *http.Request
+	workerRequest *http.Request
+	backoff       *exponentialBackoff
+}
+
+func convertToWorkerThread(thread *phpThread, worker *worker) {
+	handler := &workerThread{
+		state:  thread.state,
+		thread: thread,
+		worker: worker,
+		backoff: &exponentialBackoff{
+			maxBackoff:             1 * time.Second,
+			minBackoff:             100 * time.Millisecond,
+			maxConsecutiveFailures: 6,
+		},
+	}
+	thread.handler = handler
+	thread.requestChan = make(chan *http.Request)
+	worker.threadMutex.Lock()
+	worker.threads = append(worker.threads, thread)
+	worker.threadMutex.Unlock()
+
+	thread.state.set(stateActive)
+}
+
+func (handler *workerThread) getActiveRequest() *http.Request {
+	if handler.workerRequest != nil {
+		return handler.workerRequest
+	}
+
+	return handler.fakeRequest
+}
+
+func (t *workerThread) isReadyToTransition() bool {
+	return false
+}
+
+// return the name of the script or an empty string if no script should be executed
+func (handler *workerThread) beforeScriptExecution() string {
+	currentState := handler.state.get()
+	switch currentState {
+	case stateInactive:
+		handler.state.waitFor(stateActive, stateShuttingDown)
+		return handler.beforeScriptExecution()
+	case stateShuttingDown:
+		return ""
+	case stateRestarting:
+		handler.state.set(stateYielding)
+		handler.state.waitFor(stateReady, stateShuttingDown)
+		return handler.beforeScriptExecution()
+	case stateReady, stateActive:
+		setUpWorkerScript(handler, handler.worker)
+		return handler.worker.fileName
+	}
+	// TODO: panic?
```

Comment:
We should at least log a warning message here, maybe something like: "Unknown thread state detected, shutting down thread".

Reply:
👍  I think panicking here should be fine